Add criteria API to make it easier to set acceptance criteria for drag and drop

wbadam opened the pull request:
Fixes #9600

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/9605)
<!-- Reviewable:end -->

wbadam committed:
Add criteria API to make it easier to set acceptance criteria for drag and drop

wbadam commented:
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/9605#-Knl3Y74apiVPCSiHsT6:-Knl3Y75_JMrMNyu2Jjf:b-udcx1l) (no related file):*
Needs JavaDocs and some comments, will add in the morning

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 3 of 7 files at r1.
Review status: 3 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java, line 350 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmGJlF6ffCKmMHpyBX:-KnmGJlF6ffCKmMHpyBY:bk4yqin) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java#L350)):*
> ```Java
>                     switch (getState().criterion.getValueType()) {
>                     case Criterion.VALUE_TYPE_STRING:
>                         allowed = getState().criterion.getValue()
> ```

Why only equality comparison for strings?
Internally the operator is equals, so why not use it on the client side for symmetry?

---

*[server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java, line 297 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmGeD-B0HZ8MZhbjpm:-KnmGeD-B0HZ8MZhbjpn:b-lz12va) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java#L297)):*
> ```Java
>     }
> 
>     public void setPayload(String key, String value) {
> ```

(javadoc for these methods, related mention in documentation)

---

*[server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java, line 193 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmGuESC-xVdbAh2dpq:-KnmGuETJWtUsO5zsCJG:bcbopko) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java#L193)):*
> ```Java
>     }
> 
>     public void setDropCriteria(String key, String value) {
> ```

(javadoc for these methods)

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 1 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmFgbZ60I-MnZSSlmo:-KnmFgbZ60I-MnZSSlmp:bryekcr) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L1)):*
> ```Java
> package com.vaadin.shared.ui.dnd;
> ```

missing license header

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 5 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmHKk92tE25Z05L02f:-KnmHKk92tE25Z05L02g:bjehfje) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L5)):*
> ```Java
> import java.io.Serializable;
> 
> public class Criterion implements Serializable {
> ```

javadoc for class and methods, @since for class

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/CriterionOperator.java, line 1 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmFic7Hu2d-0CnlYsT:-KnmFic7Hu2d-0CnlYsU:bryekcr) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/shared/src/main/java/com/vaadin/shared/ui/dnd/CriterionOperator.java#L1)):*
> ```Java
> package com.vaadin.shared.ui.dnd;
> ```

missing license header

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/CriterionOperator.java, line 3 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmHO2T4wTMCBfsrABr:-KnmHO2T4wTMCBfsrABs:belnz6x) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/shared/src/main/java/com/vaadin/shared/ui/dnd/CriterionOperator.java#L3)):*
> ```Java
> package com.vaadin.shared.ui.dnd;
> 
> public enum CriterionOperator {
> ```

javadoc for class and methods, @since

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

wbadam committed:
Use compare for string values as well

wbadam committed:
Add javadoc and license

wbadam commented:
Review status: 1 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java, line 350 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmGJlF6ffCKmMHpyBX:-KnmXc8a2dE0vuouuqbk:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/client/src/main/java/com/vaadin/client/extensions/DropTargetExtensionConnector.java#L350)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Why only equality comparison for strings?
Internally the operator is equals, so why not use it on the client side for symmetry?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java, line 297 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmGeD-B0HZ8MZhbjpm:-KnmsvgBNaMbZVWA7EJl:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java#L297)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(javadoc for these methods, related mention in documentation)
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java, line 193 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmGuESC-xVdbAh2dpq:-KnmnYFmjvBmCnZ1sK8E:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java#L193)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(javadoc for these methods)
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 1 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmFgbZ60I-MnZSSlmo:-KnmZZ01B6mp1C4_Sd_c:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L1)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

missing license header
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 5 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmHKk92tE25Z05L02f:-Knmfc8nZ9ey86R_yOV5:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L5)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

javadoc for class and methods, @since for class
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/CriterionOperator.java, line 1 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmFic7Hu2d-0CnlYsT:-KnmZas3O-YWRSInNhbm:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/shared/src/main/java/com/vaadin/shared/ui/dnd/CriterionOperator.java#L1)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

missing license header
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/CriterionOperator.java, line 3 at r1](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnmHO2T4wTMCBfsrABr:-Knmi3TgVElFjoXG75jg:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/0aea201b73f8d377b4d5b15861574692dea77053/shared/src/main/java/com/vaadin/shared/ui/dnd/CriterionOperator.java#L3)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

javadoc for class and methods, @since
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

wbadam committed:
Add not equals operator

wbadam committed:
Change name of criteria script state parameter

hesara commented:
Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java, line 299 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnPfiFpCghRE-Mb4oN:-KnnPfiFpCghRE-Mb4oO:b-8tg47i) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java#L299)):*
> ```Java
> 
>     /**
>      * Sets payload for this drag source to use with acceptance criterion. The
> ```

Can you have multiple payload values for a drop target extension?

---

*[server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java, line 199 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnOHlzV655hNryfFyC:-KnnOHlzV655hNryfFyD:b-ejp4i3) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java#L199)):*
> ```Java
>      * drag source's payload with the same key. Drop is allowed if the value of
>      * the payload and the value given here are equal.
>      *
> ```

perhaps should mention what happens if this is called multiple times, and how does this relate to JS drop criteria

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 55 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnN1UxtPeqqqM5MXQO:-KnnN1UxtPeqqqM5MXQP:b-2pfjt5) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L55)):*
> ```Java
> 
>     /**
>      * Mandatory zero param constructor.
> ```

why is this required?

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 65 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnNdydo2sw8FZFecgG:-KnnNdydo2sw8FZFecgH:b-80ff7d) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L65)):*
> ```Java
>      *
>      * @param key
>      *         key of the payload to be compared
> ```

(as a general note, "payload" as a concept (in the whole class) is not necessarily clear here - I wonder if this should somehow refer to DragSourceExtension.setPayload())

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 85 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnO1yuzkSNi-haHg9i:-KnnO1yuzkSNi-haHg9j:bxbb4aj) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L85)):*
> ```Java
>      *         type of the payload to be compared
>      * @param operator
>      *         compare operator
> ```

(maybe "comparison operator")

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 172 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnNNK_hi40hYqgNjor:-KnnNNK_hi40hYqgNjos:by9bg1h) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L172)):*
> ```Java
> 
>     /**
>      * Gets the beginning of the payload as stored in data transfer object as
> ```

What does "beginning" mean? Also, the sentence is hard to read with the two "as" leading to unclear structure.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 182 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnNY57JNESsIma9aIl:-KnnNY57JNESsIma9aIm:b7s6l0w) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L182)):*
> ```Java
> 
>     /**
>      * Gets the complete payload as stored in data transfer object as data type
> ```

The sentence is hard to read with the two "as" leading to unclear structure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

wbadam committed:
Add more description to javadoc

wbadam commented:
Review status: 2 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java, line 299 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnPfiFpCghRE-Mb4oN:-KnpR4wvn5FtPe9P83KC:b-epngfc) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java#L299)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Can you have multiple payload values for a drop target extension?
</blockquote></details>

That's a good catch, I think it should't be possible to set multiple values.
I'm considering introducing a `Payload` object that would easily solve this problem. Could that be left for a next iteration?

---

*[server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java, line 199 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnOHlzV655hNryfFyC:-KnpUbOQ4WOfrGxW58DP:b-w57t1k) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java#L199)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

perhaps should mention what happens if this is called multiple times, and how does this relate to JS drop criteria
</blockquote></details>

Added a comment.
Supporting multiple criteria is a desired feature though, I just don't like any of the solutions I have in mind yet. Could that be implemented in a next iteration very soon?

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 55 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnN1UxtPeqqqM5MXQO:-KnpVPDBbr3YyFnmwBNp:b-fx3k1k) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L55)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

why is this required?
</blockquote></details>

Widgetset compiler complains about missing required (zero argument) constructor otherwise.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 65 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnNdydo2sw8FZFecgG:-KnpfI8nk9z2zarDmec1:bj2wxs5) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L65)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(as a general note, "payload" as a concept (in the whole class) is not necessarily clear here - I wonder if this should somehow refer to DragSourceExtension.setPayload())
</blockquote></details>

Added some sort of description to the class's javadoc. @link or @see is not possible due to the class being in shared module.
This whole concept should become clearer with the mentioned `Payload` object

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 85 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnO1yuzkSNi-haHg9i:-KnpVvxKiRgiEoZU2THo:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L85)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(maybe "comparison operator")
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 172 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnNNK_hi40hYqgNjor:-Knpf7vTq7_2bwlsFbF7:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L172)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

What does "beginning" mean? Also, the sentence is hard to read with the two "as" leading to unclear structure.
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java, line 182 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnNY57JNESsIma9aIl:-Knpf8_gecpoexIDfSpE:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/shared/src/main/java/com/vaadin/shared/ui/dnd/Criterion.java#L182)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

The sentence is hard to read with the two "as" leading to unclear structure.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java, line 199 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnOHlzV655hNryfFyC:-KnramSh63A7EZCMz-ZH:by3mpia) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java#L199)):*
<details><summary><i>Previously, wbadam (Adam Wagner) wrote…</i></summary><blockquote>

Added a comment.
Supporting multiple criteria is a desired feature though, I just don't like any of the solutions I have in mind yet. Could that be implemented in a next iteration very soon?
</blockquote></details>

we are trying to get 8.1 out ASAP, so an iteration that misses the last cut-off should then have a backwards compatible API (e.g. add a new method addDropCriterion(...))

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

hesara commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java, line 299 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnPfiFpCghRE-Mb4oN:-KnrbBJaOZ7pqYFIdbvl:b1mv1pb) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java#L299)):*
<details><summary><i>Previously, wbadam (Adam Wagner) wrote…</i></summary><blockquote>

That's a good catch, I think it should't be possible to set multiple values.
I'm considering introducing a `Payload` object that would easily solve this problem. Could that be left for a next iteration?
</blockquote></details>

when would that iteration be, as we are very close to 8.1 RC?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

wbadam committed:
Introduce Payload object and refactor criteria comparison

hesara commented:
Reviewed 10 of 12 files at r4.
Review status: 9 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 417 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-Knsk_1A2Ua4XgHJj5g3:-Knsk_1A2Ua4XgHJj5g4:b9gkxti) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L417)):*
> ```Java
> 
>         // Add payload for comparing against acceptance criteria
>         getState().payload.values().forEach(payload -> orderedData
> ```

(does this approach work properly also in IE11?)

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 418 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-Knsk3Bey58jJYjpLNH6:-Knsk3Bey58jJYjpLNH7:bigyzy5) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L418)):*
> ```Java
>         // Add payload for comparing against acceptance criteria
>         getState().payload.values().forEach(payload -> orderedData
>                 .put(payload.toString(), payload.getValue()));
> ```

I would prefer not relying on the generic toString() but rather some other, better named method

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java, line 100 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnskiWMZcL4fKndS2Bu:-KnskiWMZcL4fKndS2Bv:b-4wjm04) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java#L100)):*
> ```Java
>      */
>     @Override
>     public String toString() {
> ```

see comment above - would prefer calling a different named method over an override of toString() for this purpose (and optionally toString() calling that method if that is wanted for debug purposes)

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java, line 122 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnslQDGjY0InwCUrSOm:-KnslQDGjY0InwCUrSOn:bgc9a22) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java#L122)):*
> ```Java
> 
>         return new Payload(parts[2], parts[3],
>                 ValueType.valueOf(parts[1].toUpperCase()));
> ```

why toUpperCase() here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

wbadam committed:
Add option to set multiple criteria and tell whether any or all should match

wbadam committed:
Rename payload toString method

wbadam commented:
Review status: 9 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 417 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-Knsk_1A2Ua4XgHJj5g3:-KnsvwgSj7IVP6ptKc1J:b-uzlqpj) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L417)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(does this approach work properly also in IE11?)
</blockquote></details>

Unfortunately the whole feature doesn't work on IE since it doesn't support any data types other than text. Pekka's opinion is that it should be stated in the docs that this will not work on that browser.

Another option I considered is to store the payload manually on the client but that would only work within the same UI.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 418 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-Knsk3Bey58jJYjpLNH6:-KnsxLdoqAqB2Sd_Hna-:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L418)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

I would prefer not relying on the generic toString() but rather some other, better named method
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java, line 299 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnPfiFpCghRE-Mb4oN:-KnswgeDPZz9c0tSd5yL:b-tagqb0) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/server/src/main/java/com/vaadin/ui/dnd/DragSourceExtension.java#L299)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

when would that iteration be, as we are very close to 8.1 RC?
</blockquote></details>

It's implemented now

---

*[server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java, line 199 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnOHlzV655hNryfFyC:-KnswlL4LVnFyNBiLuC9:b-tagqb0) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java#L199)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

we are trying to get 8.1 out ASAP, so an iteration that misses the last cut-off should then have a backwards compatible API (e.g. add a new method addDropCriterion(...))
</blockquote></details>

It's implemented now

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java, line 100 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnskiWMZcL4fKndS2Bu:-KnsxEWFbEaL7ZZE3uh_:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java#L100)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

see comment above - would prefer calling a different named method over an override of toString() for this purpose (and optionally toString() calling that method if that is wanted for debug purposes)
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java, line 122 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnslQDGjY0InwCUrSOm:-KnsxlDswhsLEp43y04o:b-vt7g5l) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java#L122)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

why toUpperCase() here?
</blockquote></details>

Browser will convert data type to lower case and enum is has upper case value.
Could make enum's values lower case if you think it's a nicer solution?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 417 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-Knsk_1A2Ua4XgHJj5g3:-Ko6Dr2Jd1n2ZCdCI31w:b3t8k7m) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L417)):*
<details><summary><i>Previously, wbadam (Adam Wagner) wrote…</i></summary><blockquote>

Unfortunately the whole feature doesn't work on IE since it doesn't support any data types other than text. Pekka's opinion is that it should be stated in the docs that this will not work on that browser.

Another option I considered is to store the payload manually on the client but that would only work within the same UI.
</blockquote></details>

probably should mention the limitation in related javadocs, then

---

*[server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java, line 199 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnOHlzV655hNryfFyC:-Ko6DG9lTotNgFQlRLyS:b-6301m7) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java#L199)):*
<details><summary><i>Previously, wbadam (Adam Wagner) wrote…</i></summary><blockquote>

It's implemented now
</blockquote></details>

The comment here and on the other variants about only one criterion being supported is no longer quite accurate, and could also mention relationship to JS criteria and note that more complex criteria can be handled with it (e.g. JS does not overwrite these criteria and vice versa).

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/DropTargetState.java, line 42 at r5](https://reviewable.io:443/reviews/vaadin/framework/9605#-Ko6EFlNFMx5ezvv6kM-:-Ko6EFlNFMx5ezvv6kM0:b-54wmux) ([raw file](https://github.com/vaadin/framework/blob/258c7e7a4ff3bb5e803d2ebe7342d8e5b37e3ec9/shared/src/main/java/com/vaadin/shared/ui/dnd/DropTargetState.java#L42)):*
> ```Java
> 
>     /**
>      * List of criterion to compare against the payload.
> ```

(criteria)

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Criterion.java, line 117 at r5](https://reviewable.io:443/reviews/vaadin/framework/9605#-Ko6HVXxMV3W9sLc_t2C:-Ko6HVXxMV3W9sLc_t2D:bqmykwv) ([raw file](https://github.com/vaadin/framework/blob/258c7e7a4ff3bb5e803d2ebe7342d8e5b37e3ec9/shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Criterion.java#L117)):*
> ```Java
>      *         comparison operator
>      */
>     private Criterion(String key, String value,
> ```

different parameter order than for other variants

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Criterion.java, line 168 at r5](https://reviewable.io:443/reviews/vaadin/framework/9605#-Ko6HhCkGq8lBvgoVqkB:-Ko6HhCkGq8lBvgoVqkC:b7p7axl) ([raw file](https://github.com/vaadin/framework/blob/258c7e7a4ff3bb5e803d2ebe7342d8e5b37e3ec9/shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Criterion.java#L168)):*
> ```Java
>      *         collection of payloads to compare the criterion against
>      * @return {@code false} if there exists a payload in the collection with
>      * the same key and value type and it doesn't match the criterion, {@code
> ```

that payloads of other same name but different type are ignored in the comparison is somewhat non-obvious, so could perhaps try to make that clearer also in the comment above

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java, line 122 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnslQDGjY0InwCUrSOm:-Ko6ER0lRfHoywl5GvU-:b-n6xh3w) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java#L122)):*
<details><summary><i>Previously, wbadam (Adam Wagner) wrote…</i></summary><blockquote>

Browser will convert data type to lower case and enum is has upper case value.
Could make enum's values lower case if you think it's a nicer solution?
</blockquote></details>

this is ok, though it would not hurt to have a comment about this

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

wbadam committed:
Add javadoc clarifications

wbadam commented:
Review status: 5 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 417 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-Knsk_1A2Ua4XgHJj5g3:-Ko6to9D7hzgEDbKLuj5:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L417)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

probably should mention the limitation in related javadocs, then
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java, line 199 at r2](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnnOHlzV655hNryfFyC:-Ko719JNdf-0BlP6QAg0:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/6843d9b413bb2426964f1b23e2f3b91b4018f159/server/src/main/java/com/vaadin/ui/dnd/DropTargetExtension.java#L199)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

The comment here and on the other variants about only one criterion being supported is no longer quite accurate, and could also mention relationship to JS criteria and note that more complex criteria can be handled with it (e.g. JS does not overwrite these criteria and vice versa).
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/DropTargetState.java, line 42 at r5](https://reviewable.io:443/reviews/vaadin/framework/9605#-Ko6EFlNFMx5ezvv6kM-:-Ko71IetUQe10AjEAqkE:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/258c7e7a4ff3bb5e803d2ebe7342d8e5b37e3ec9/shared/src/main/java/com/vaadin/shared/ui/dnd/DropTargetState.java#L42)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(criteria)
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Criterion.java, line 117 at r5](https://reviewable.io:443/reviews/vaadin/framework/9605#-Ko6HVXxMV3W9sLc_t2C:-Ko71ZwDbnJoLreZOm2n:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/258c7e7a4ff3bb5e803d2ebe7342d8e5b37e3ec9/shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Criterion.java#L117)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

different parameter order than for other variants
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Criterion.java, line 168 at r5](https://reviewable.io:443/reviews/vaadin/framework/9605#-Ko6HhCkGq8lBvgoVqkB:-Ko72Yj3XozOtLHMumUQ:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/258c7e7a4ff3bb5e803d2ebe7342d8e5b37e3ec9/shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Criterion.java#L168)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

that payloads of other same name but different type are ignored in the comparison is somewhat non-obvious, so could perhaps try to make that clearer also in the comment above
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java, line 122 at r4](https://reviewable.io:443/reviews/vaadin/framework/9605#-KnslQDGjY0InwCUrSOm:-Ko737i381jGoRdfODIU:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/acbe5f02b0b3a58e68306832c7d3b368674e0ea4/shared/src/main/java/com/vaadin/shared/ui/dnd/criteria/Payload.java#L122)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

this is ok, though it would not hurt to have a comment about this
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 6 of 6 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9605)*
<!-- Sent from Reviewable.io -->
